software/control_sw/src/cosmic_fengine.py

```diff
@@ -1437,36 +1447,54 @@ def delay_tracking(self):
 
             else:
                 #No values received, load values on hand
-                required_loadtime_s = int(delay_coeffs['loadtime']) #assuming in integer seconds
-                required_loadtime_fpga_clks = required_loadtime_s * FPGA_CLOCK_RATE_HZ
-                
+                t = time.time() * 1e6 # received loadtime is in microseconds so compare against time now in microseconds
+
+                if (delay_coeffs['loadtime_us'] < (t + 1e5)
+                    or 
+                    delay_coeffs['loadtime_us'] > (t + 1e6)):
+                    #if the provided loadtime is not at least 1e-1s into the future or more than 1s into the future
+                    required_loadtime_us = t + 1e6 # interpolate one second into the future on the old delay coefficients
+                    required_loadtime_s = required_loadtime_us * 1e-6
+                else:
+                    #if the provided loadtime is in the future but within 1 second 
+                    required_loadtime_us = int(delay_coeffs['loadtime'])
+                    required_loadtime_s = required_loadtime_us * 1e-6
+                    #update values
+                    delay_ns = delay_coeffs["delay_ns"]
+                    delay_rate_nsps = delay_coeffs["delay_rate_nsps"]
+                    delay_raterate_nsps2 = delay_coeffs["delay_raterate_nsps2"]
+                    sideband_0 = delay_coeffs["sideband_0"]
+                    sideband_1 = delay_coeffs["sideband_1"]
+                    sslo_0 = delay_coeffs["effective_lo_0_mhz"]
+                    sslo_1 = delay_coeffs["effective_lo_1_mhz"]
+                    fshifts = np.array(delay_coeffs["lo_hz"],dtype=float)                                           #fshift in Hz
+
+                required_loadtime_fpga_clks = int((required_loadtime_us * FPGA_CLOCK_RATE_HZ) * 1e-6) #should give us fpga_clks in microsecond precision
```

Comment:
I think I would feel happier if this was `(required_loadtime_us * FPGA_CLOCK_RATE_HZ) // 1000000` but maybe I'm just paranoid.
Also, I think delays in the phase tracker probably have to be loaded on spectra boundaries, so if the resultant value here isn't an appropriate multiple of `4096*ADC_CLOCK_RATE` (`=512*FPGA_CLOCK_RATE`) something bad might happen. I'm not sure if this is enforced by the loading method

Reply:
@jack-h 
I'll do the integer division but I don't think we assert that the load must be a multiple of `512*FPGA_CLOCK_RATE`. Underlying loadtime function for the `phaserotate` and `delay` just call [this](https://github.com/realtimeradio/vla-dev/blob/master/software/control_sw/src/blocks/timed_pulse.py#L48) function... Shall I try enforce it with something like:
`assert required_loadtime_fpga_clks%512*FPGA_CLOCK_RATE == 0.0`?

Reply:
This of course does put restrictions upstream on the loadtime the automator can request?

Reply:
I think this is a good idea so we at least see the issue.  I think it's probably reasonable to only allow upstream requests to be on spectrum boundaries.
Ps. That assert needs brackets around (512*...) and floats are still evil

Reply:
Decided unecessary